### ISSUE TITLE
Elevated custom managed dialog before Dialogs.Progress

### ISSUE THREAD

**Issue by DarthSidius** [oleg-shilo/wixsharp#384]
Subj.
Or the whole setup process elevated

**Comment by oleg-shilo** [comment]
It is problematic to elevate a specific dialogs. The whole dialog sequence is a single CA that MSI launches unelevated. 

It's tempting to try to find a work around to fool MSI (EmbeddedUI functionality) to launch it elevated. However the chances that this can be done are slim. And also it may not be reliable even if the work around is found.

I would encourage you instead of trying to force MSI to do something that it "doesn't like" use standard approach, promoted by MS just after the introduction of UAC in Vista. This approach was a part of the requirements for an application to be certified as "Vista compatible".

The approach is in fact very simple and reliable. The main process (e.g. my_app_config.exe) is never elevated. This creates beneficial user experience (e.g. quick startup, no user interaction at startup). However when an elevated action is required a special action utility is started as an elevated background process. This utility performs the action and unelevated UI simply waits for that utility process to succeed or to fail. 

BTW MSI uses the same approach. Every single Custom action is launched as a separate process rundll32.exe. Thus nothing new. You can just build and pack such a utility into your msi and then launch it before the progress dialog.

**Comment by DarthSidius** [comment]
It turns out that only `AftrerInstall` event runs elevated (

**Comment by oleg-shilo** [comment]
Yes it is. the case. The GUI supposed to run unelevated (regardless if you are using WixSharp or not). 

The setup **_standard actions_** elevation happens only after you start the actual installation (after progress  dialog is popped up). The **_custom actions_** are always unelevated unless you it is a special type of action - "deferred" (e.g. AfterInstall event).  

Though I am not sure why it is relevant. Your thread title "_Elevated custom managed dialog_" clearly indicated that you are not looking for the elevated custom action but a **dialog**. 

If you want to elevate a custom action then you may try to use `ElevatedCustomAction` (DeferredAction sample):
``` C#
var project =
    new ManagedProject("My Product",
        new Dir(@"%ProgramFiles%\My Company\My Product",
            new File(@"Files\MyApp.exe"),
            new File(@"Files\MyApp.exe.config",
                new XmlFile(XmlFileAction.setValue, @"//configuration/appSettings/add[\[]@key='AppName'[\]]/@value", "My App"),
                new XmlFile(XmlFileAction.setValue, @"//configuration/connectionStrings/add[\[]@name='Server1'[\]]/@connectionString", "DataSource=(localdb)/v11.0;IntegratedSecurity=true"),
                new XmlFile(XmlFileAction.setValue, @"//configuration/connectionStrings/add[\[]@name='Server1'[\]]/@providerName", "System.Data.SqlClient"))),
        new ElevatedManagedAction(CustomActions.OnInstall, Return.check, When.After, Step.InstallFiles, Condition.NOT_Installed)
        {
            UsesProperties = "CONFIG_FILE=[INSTALLDIR]MyApp.exe.config, APP_FILE=[INSTALLDIR]MyApp.exe, DATABASE_CONNECTION_STRING=[DATABASE_CONNECTION_STRING]",
            Execute = WixSharp.Execute.deferred
        });
```

**Comment by DarthSidius** [comment]
I apologize for my English - I use Google Translate.
I thought that you could just insert your own into the sequence of dialogues. Like this:
```
	proj.ManagedUI = new ManagedUI();
	proj.ManagedUI.InstallDialogs
		.Add(Dialogs.Welcome)
		.Add(Dialogs.InstallDir)
		.Add<MyCustomDialog>()
		.Add(Dialogs.Progress)
		.Add(Dialogs.Exit)
	;
```
There is another option to run the entire installation package elevated. Elevated bootstrapper? But the example of the Elevated bootstrapper did not understand.
I know about manifestos. But if it does not complicate you, show a simple example of bootraper, whose task is only to run the installation package elevated.

**Comment by oleg-shilo** [comment]
Not a problem. 
You can insert the dialog as per your code snippet but it will become just a step within a sequence of dialogs and that whole sequence is a dedicated custom action which only MSI has a full control over.

Elevating the whole UI is a challenge and bootstrapper is the most direct way of solving it. You can for example use SilentBA sample from the WixSharp samples. 

I think your task has a much simpler solution. You can subscribe for an early UI step event (e.g. UIInitialized for managed UI). Then check if you are elevated and restart setup if you are not:

```C#
static void Project_UIInitialized(SetupEventArgs e)
{
    if (!new WindowsPrincipal(WindowsIdentity.GetCurrent()).IsInRole(WindowsBuiltInRole.Administrator))
    {
        MessageBox.Show(e.Session.GetMainWindow(), "You must start the msi file as admin");
        e.Result = ActionResult.Failure;

        var startInfo = new ProcessStartInfo();
        startInfo.UseShellExecute = true;
        startInfo.WorkingDirectory = Environment.CurrentDirectory;
        startInfo.FileName = "msiexec.exe";
        startInfo.Arguments = "/i \"" + e.MsiFile + "\"";
        startInfo.Verb = "runas";

        Process.Start(startInfo);
    }
}
```  

I just tested it with the SetupEvents sample and works. It restarts itself quite smoothly.

**Comment by oleg-shilo** [comment]
I have just committed the sample that restarts the msi elevated if it is not. YOur use case is practically the same except you are testing the system locale instead of the user privileges:
https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/RestartElevated(UI)
